### PR TITLE
13 allow numpy inputs clean

### PR DESCRIPTION
**Pull request by Alastair-L**
Clean implementation of: https://github.com/marioboley/realkd.py/pull/2/files

Card: https://trello.com/c/cPmxux7S/13-allow-numpy-inputs-to-fit-and-predict

### CODE REVIEW

File: realkd/rules.py [hunk 1]
```diff
@@ -167,6 +168,10 @@ class Rule:
     >>> r = Rule(female, 1.0, 0.0)
     >>> r(titanic.iloc[0]), r(titanic.iloc[1])
     (0.0, 1.0)
+    >>> r(titanic[titanic.index == 0]), r(titanic[titanic.index == 1])
```
Comment (Alastair-L) [inline realkd/rules.py:171]: REGRESSION:
Rules (and conjunctions etc.), when passed a dataframe with one row will return a one element series, not a float.

This makes some sense to me, but wanted to flag it as a regression. Not sure what the standard is here.

File: realkd/rules.py [hunk 2]
```diff
@@ -412,14 +420,14 @@ def bound(self, ext):
     def opt_weight(self, q):
         # TODO: this should probably just be defined for ext (saving the q evaluation)
         # ext = self.ext(q)
-        ext = self.data.loc[q].index
+        ext = q(self.data).nonzero()[0]
```
Comment (Alastair-L) [inline realkd/rules.py:423]: Since this is a common pattern and in my opinion quite hard to read - I think it deserves a function - but I was trying to be as minimal as possible so didn't add anything.

File: realkd/search.py [hunk 3]
```diff
@@ -189,31 +191,62 @@ def from_df(df, without=None, max_col_attr=10, sort_attributes=True, discretizat
         >>> titanic_ctx.m
         891
         >>> titanic_ctx.attributes # doctest: +NORMALIZE_WHITESPACE
-        [Survived<=0, Survived>=1, Pclass<=1, Pclass<=2, Pclass>=2, Pclass>=3, Sex==male, Sex==female, Age<=23.0,
+        [Survived<=0, Survived>=1, Pclass<=1, Pclass<=2, Pclass>=2, Pclass>=3, Sex==female, Sex==male, Age<=23.0,
```
Comment (Alastair-L) [inline realkd/search.py:194]: REGRESSION:
Attributes created here will now be ordered alphabetically. I think this is a positive change

File: realkd/subgroups.py [hunk 4]
```diff
@@ -9,6 +9,7 @@
 
 from realkd.search import Conjunction, Context, KeyValueProposition, Constraint, search_methods
 from realkd.rules import Rule
+from realkd.utils import validate_data
 
 
 class Impact:
```
Comment (Alastair-L) [inline realkd/subgroups.py:15]: NOTE:
Although there is no regression in Impact, I did not end up supporting the numpy array input because 
a) I ran out of time
b) I'm trying to be super conservative with changes.

File: realkd/subgroups.py [hunk 5]
```diff
@@ -33,20 +34,21 @@ class Impact:
...
         self.m = len(data)
-        self.data = data.sort_values(target, ascending=False)  # data
+        self.data = data[argsort(data[target])[::-1]]
         self.data.reset_index(drop=True, inplace=True)
         self.target = target
         self.mean = self.data[self.target].mean()
 
     def __call__(self, q):
-        extent = self.data.loc[q]
+        extent = self.data[q]
```
Comment (Alastair-L) [inline realkd/subgroups.py:46]: NOTE:
This is called both with a query (callable) and a list of extents, hence why the q(self.data) pattern was not used here and also why numpy input is not supported for Impact

File: realkd/utils.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+def validate_data(data, labels=None):
+    return data if isinstance(data, RealkdArrayLike) else RealkdArrayLike(data, labels)
+
+class RealkdArrayLike:
+    @staticmethod
+    def get_labels(data, labels=None):
+        if labels is not None:
+            return labels
+        elif hasattr(data, "columns"):
+            return data.columns.to_list()
```
Comment (Alastair-L) [inline realkd/utils.py:46]: I probably could/should have embraced the duck typing and just set labels to data.columns, since it's iterable, however I do use the list-only method `index`. Do you think I made the right call here?

image

File: realkd/utils.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                # If the array is 1D, treat it as an observation
+                # If the array is 2D, treat it the first dimention (rows) as observations
+                if self._raw.ndim == 1:
+                    return self._raw[self.labels.index(first_key)], [first_key]
+                return self._raw[:, self.labels.index(first_key)], [first_key]
+            return self._raw.__getitem__(first_key), [first_key]
+
+        else:
+            if hasattr(self._raw, "iloc"):
+                try:
```
Comment (Alastair-L) [inline realkd/utils.py:116]: This seems really hacky - do you have a cleaner solution for handling a range of slices, callables, arraylikes and integers?

File: realkd/utils.py [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+        """
+        return self._raw.__eq__(other)
+
+    def __gt__(self, other): return self._raw.__gt__(other)
+    def __ge__(self, other): return self._raw.__ge__(other)
+    def __lt__(self, other): return self._raw.__lt__(other)
+    def __le__(self, other): return self._raw.__le__(other)
+    def __neg__(self): return self._raw.__neg__()
+
+    def __getattr__(self, name):
```
Comment (Alastair-L) [inline realkd/utils.py:147]: Do you think this is a good idea or is it **too** permissive?

I thought a potential middle ground could be to add a 
print("Warning - this may be 100% supported") before attribute access so that our own code doesn't accidentally use it and couple itself to a particular data structure? What do you think?

**Review by Alastair-L — COMMENTED** [review]
Hi @marioboley!

I'd love your review on this code - I'd love to set up a meeting at some point if that would be more convenient for you?

File: realkd/rules.py (hunk 1 above)
Comment (marioboley) [reply realkd/rules.py:171]: Can you please check the behaviour of sklearn for this question? On first glance it seems that there should be shape matching in the following sense:

shape 1 x n -> shape 1 x 1

Only the changed behaviour would satisfy that if I understand you correctly. The one-row/one-column input do generally require some work from the caller also in sklearn. If you just input a 1d-array you are asked to convert it to a matrix.

File: realkd/rules.py (hunk 2 above)
Comment (marioboley) [reply realkd/rules.py:423]: I think it is actually quite readable. The only unexpected part is the final `[0]` , which I assume is necessary because numpy returns a matrix instead of a vector?

BTW: The todo comment points to a straightforward performance improvement for this specific function

File: realkd/search.py (hunk 3 above)
Comment (marioboley) [reply realkd/search.py:194]: Why is this a positive change? Why was it necessary/convenient?

File: realkd/subgroups.py (hunk 5 above)
Comment (marioboley) [reply realkd/subgroups.py:46]: Ok, this is interesting. It means that one functionality of pandas that was used internally is indexing with a callable. Can we come up with a general refactoring for these cases? I suppose the straightforward solution is to have 

```
self.data[q(self.data)]
```

Does that look good to you?

Comment (Alastair-L) [reply realkd/subgroups.py:46]: I don't believe that will work out of the box, since q could not be callable. Your solution looks good if q is guaranteed to be callable but here it might not be.
A solution could be:
```
if callable(q):
  self.data[q(self.data)]
else:
  self.data[q]
```
